rustler/src/types/atom.rs

```diff
@@ -99,6 +100,12 @@ impl<'a> Decoder<'a> for Atom {
     }
 }
 
+impl Hash for Atom {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_c_arg().hash(state);
```

Comment:
Is this guaranteed to work? Wouldn't it be better to use `enif_hash` instead?

Reply:
I think this works if we can rely on the NIF_TERM of the same atom being stable across several "instances" of the same atom, right? We are relying on that property for equality checking of atoms as well I believe, so this should hold up here as well. Am I missing something?

But using enif_hash sounds reasonable, as that function knows more about the type.

Reply:
The only reason why this would work is that Erlang "inlines" atoms:

https://github.com/erlang/otp/blob/0a3e25826ccbc160cd1dd3866fc3457b9241d8fb/erts/emulator/beam/erl_term.h#L293

As said, I regard this as an implementation detail. There have been multiple discussions over time to loosen the finite/non-gc'd atom table restriction, and if that actually materialized, code relying on the stability of `Eterm` objects would be problematic.

Reply:
Sounds reasonable. But will we need to rethink equality as well then? We currently just derive it.